ons-spark/spark-analysis/Cramer_v.md

```diff
@@ -0,0 +1,439 @@
+```
+
+```{code-tab} r R
+
+get_cramer_v(freq_r)
+
+```
+````
+
+````{tabs}
+
+```{code-tab} plaintext Python Output
+0.086908868435997877
+```
+
+```{code-tab} plaintext R Output
+[1] 0.08690887
+```
+````
+We now get an Cramér V of 0.0869, telling us that there is little to no association between `cal_year` and `animal_type`.
+This example also validates both Python and R methods against each other, as we get identical outputs.
+
+#### Cramér's V Example 2: `postcode_district` and `animal_type`
+
+Using the earlier defined function, just need to repeat the cross tabbing exercise before applying our function. This time we will select `postcode_district` instead of `cal_year`.
+````{tabs}
+```{code-tab} py
+freq_spark = rescue.crosstab('postcode_district','animal_type')
+freq_pandas = freq_spark.toPandas()
+freq_pandas_reindexed = freq_pandas.set_index('postcode_district_animal_type')
+freq_numpy = np.array(freq_pandas_reindexed)
+get_cramer_v(freq_numpy)
+```
+
+```{code-tab} r R
+
+freq_spark <- sdf_crosstab(rescue, 'postcode_district', 'animal_type') 
+freq_r <- freq_spark %>% collect() 
+freq_r <- subset(freq_r, select = -postcode_district_animal_type)
+get_cramer_v(freq_r)
+
+
+```
+````
+
+````{tabs}
+
+```{code-tab} plaintext Python Output
+0.29083878195595769
+```
+
+```{code-tab} plaintext R Output
+[1] 0.2908388
+```
+````
+This time we get a cramér V value of 0.29, suggesting a slight association between `postcode_district` and `animal_type`.
+
+### Potential Issue with `sdf_crosstab()`
+During the testing for this page, we noted a few common issues which may arise when attempting to calculate Cramér's V statistic. Specifically this issue is linked to `sdf_crosstab()` function for data which has not been fully pre processed.
+
+For this example we take the unprocessed rescue dataset. Here in `animal_type`, we have values for cat with both a upper and lower case first letter. Here we will use the `tryCatch()` function to handle the error message, this is similar to pythons `try: except:` statements (For more info see [`Error handing in R`](https://cran.r-project.org/web/packages/tryCatchLog/vignettes/tryCatchLog-intro.html) or [`Errors and Exceptions (Python)`](https://docs.python.org/3/tutorial/errors.html)).
+
+
+````{tabs}
+
+```{code-tab} r R
+
+rescue_raw <- sparklyr::spark_read_csv(sc, config$rescue_path_csv, header=TRUE, infer_schema=TRUE)
+rescue_raw <- rescue_raw %>%
+    dplyr::rename(
+        incident_number = IncidentNumber,
+        animal_type = AnimalGroupParent,
+        cal_year = CalYear)
+
+tryCatch(
+  {
+    sdf_crosstab(rescue_raw, 'cal_year', 'animal_type')
+  },
+  error = function(e){
+    message('Error message from Spark')
+    print(e)
+  }
+)
+
+```
+````
+
+````{tabs}
+
+```{code-tab} plaintext R Output
+<simpleError: org.apache.spark.sql.AnalysisException: Reference 'Cat' is ambiguous, could be: Cat, Cat.;
+	at org.apache.spark.sql.catalyst.expressions.package$AttributeSeq.resolve(package.scala:259)
```

Comment:
Got a different error :
Error message from Spark
<error/rlang_error>
Error:
! org.apache.spark.sql.AnalysisException: Reference 'Cat' is ambiguous,
  could be: Cat, Cat.;

Run `sparklyr::spark_last_error()` to see the full Spark error (multiple lines)
To use the previous style of error message set
`options("sparklyr.simple.errors" = TRUE)`
---
Backtrace:
     ▆
  1. ├─base::tryCatch(...)
  2. │ └─base (local) tryCatchList(expr, classes, parentenv, handlers)
  3. │   └─base (local) tryCatchOne(expr, names, parentenv, handlers[[1L]])
  4. │     └─base (local) doTryCatch(return(expr), name, parentenv, handler)
  5. ├─sparklyr::sdf_crosstab(rescue_raw, "cal_year", "animal_type")
  6. │ └─... %>% sdf_register()
  7. ├─sparklyr::sdf_register(.)
  8. ├─sparklyr::invoke(...)
  9. └─sparklyr:::invoke.shell_jobj(...)
 10.   ├─sparklyr::invoke_method(...)
 11.   └─sparklyr:::invoke_method.spark_shell_connection(...)
 12.     └─sparklyr:::core_invoke_method(...)
 13.       └─sparklyr:::core_invoke_method_impl(...)
 14.         └─sparklyr:::spark_error(msg)